src/ui/components/shared/Tooltip.tsx

```diff
@@ -0,0 +1,137 @@
+import { useCallback, useLayoutEffect, useRef, useState } from "react";
+
+const TOOLTIP_OFFSET_BOTTOM = 10;
+const TOOLTIP_OFFSET_TOP = 5;
+
+type Coordinates = {
+  pageX: number;
+  pageY: number;
+};
+
+type Elements = {
+  container: HTMLElement | null;
+  tooltip: HTMLDivElement | null;
+};
+
+type TooltipProps = {
+  label: string;
+  targetID?: string;
+};
+
+export default function Tooltip({ label, targetID }: TooltipProps) {
+  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);
+
+  const elementsRef = useRef<Elements>({
+    container: null,
+    tooltip: null,
+  });
+
+  const tooltipRefSetter = useCallback((tooltip: HTMLDivElement | null) => {
```

Comment:
Using a ref-setter function lets us do a one-time lookup of the nearest positioned parent and cache the value for later.